### ISSUE TITLE
Guidance on rotation of certificates and client secrets

### ISSUE THREAD

**Issue by debben** [panva/node-oidc-provider#144]
I have somewhat of a two part question related to the rotation of sensitive values used by the OP. 

With regard to the keystore for id_token signing, I see the keystore is being queried just by algorithm. Node-jose's keystore implementation is written such that if multiple keys meet the criteria, the one at index 0 in the result set is returned. In this way I could introduce a new certificate pair as the first item and all subsequent id_tokens issued would be signed with the new certificate. The issue though is that when my clients go to verify, there is no "kid" in the header and my client library can't discern which key from /op/certs should be used. I know this is probably somewhat of an edge case given that my clients usually validate id_token signatures immediately after the id_token is received. I could just as well replace my certificate rather than having both in the keystore, but this overlapping technique for rotating keys is something I think should be supported. The library I use for one of my clients I believe caches the result of /op/certs for a brief amount of time and I'd want a way to have the new key out for awhile before being made the one used for signing. Is it possible to add 'kid' to the id_token headers?

I also see similar behavior for the integrity keystore that would prevent me from rotating that secret. When replacing that key or even inserting a new one, existing tokens in the keystore are unable to be verified. Having the kid in the header would allow for keys to be rotated. I see though that in 2.x integrity keystore is removed so this is somewhat irrelevant.

The second part of my question is with regard to the rotation of client_secret in the client config. In other OIDC providers such as just using Azure AD, I configure a client and am given a client id. I can then configure multiple secrets with varying expiry dates. Right now with node-oidc-provider I have several clients configured that just use the client_credentials grant to obtain a token. These clients behave like service accounts and use the tokens they obtain to be able to introspect tokens issued to other clients through other flows.

What would be great is if clients could support multiple client secrets. In this way I could create a new secret for a given client_id and give that to my client to be installed in their application. Once they confirm they've updated the secret I can remove the old one or even have this done automatically (in my database adapter for clients I separate client_secret into it's own table with columns for expiry dates). At some level I could already support this behavior just by creating a new client with all of the same configuration except the client_id and client_secret. I'd then have the client update both values. I currently use client_id in logging as well as to describe relationships between the client and other tables in my application, so having to change the client_id would be less than ideal. Is there any merit to supporting multiple client_secrets within client configs? It would require updating files like token_auth.js to try validating against every secret in the configuration rather than just the one.

**Comment by panva** [comment]
Hello @debben, thank you for reaching out.

First of all, you are right, integrity as a feature is out. No need to bother i reckon. I will eventually even rewrite the complete adapter structure to use something more "standard".

> there is no "kid" in the header

Are you sure? If I recall correctly there's always a `kid` claim in the oidc-provider ID Token JWT headers, either the one you pass on to the key or node-jose forms one using the key material. 

The rotation is then rather easy. All good client libraries detect that there's a new `kid` value they don't know and go re-fetch the `jwks_uri` contents. Obviously the key then needs to be available on all nodes, else (also my library) will not re-fetch for a while as it caches the responses for at least a minute. So, for distributed deployments I would recommend to deploy the new keys first by having them at the end of the `keys` array, this will make them available in the published keys uri but not used yet (as you said yourself, node-jose will take the first one). Once deployed and published on all hosts do another deploy which simply pushes them to the front. This will make them used and will trigger re-fetches by clients and the keys will be already available on all the hosts/processes that respond.

> What would be great is if clients could support multiple client secrets. Is there any merit to supporting multiple client_secrets within client configs?

I don't see oidc-provider supporting out of spec proprietary mechanisms. 
Eventually i would love to support expiring dynamic registrations (client_secret_expires_at) and taking over some properties from a register initial access token, but this would not solve your problem, would it?

**Comment by debben** [comment]
Thanks for the quick response @panva .

> Are you sure? If I recall correctly there's always a kid claim in the oidc-provider ID Token JWT headers, either the one you pass on to the key or node-jose forms one using the key material.

You are correct. I was mistaken. I have multiple clients some of which sign id_token with HS rather than RS. I was looking at one of the HS id_tokens when I originally wrote this and thought kid was not being included. It is being included for RS signed id_tokens which is what I would expect.

> I don't see oidc-provider supporting out of spec proprietary mechanisms.
Eventually i would love to support expiring dynamic registrations (client_secret_expires_at) and taking over some properties from a register initial access token, but this would not solve your problem, would it?

I understand and agree. I think I answered my own question to some extent in my proposal to distribute new client_ids in addition to new client_secrets. The trouble is that I had started to build around client_id in my implementation. For example I check during the interaction process that the user completing the interaction actually has access to client that initiated the flow. This is done by tables in my implementation that map user access to clients. I'll just have to refactor things to tie things to an 'application' which has one or more clients.
